File: homework01-go/vigenere/vigenere.go
package vigenere

import (
	"strings"
)

func EncryptVigenere(plaintext string, keyword string) string {
	var ciphertext string

	keyword = strings.ToLower(keyword)
	for i := 0; i < len(plaintext); i++ {
		letter := plaintext[i]
		shift := int(keyword[i%len(keyword)]) - 97
		if (int('a') <= int(letter)) && (int(letter) <= int('z')) {
			ciphertext += string(rune(int(letter)-97+shift)%26 + 97)
		} else if (int('A') <= int(letter)) && (int(letter) <= int('Z')) {
			ciphertext += string(rune(int(letter)-65+shift)%26 + 65)
		} else {
			ciphertext += string(letter)
		}
	}

	return ciphertext
}

func DecryptVigenere(ciphertext string, keyword string) string {
	var plaintext string

	keyword = strings.ToLower(keyword)
	for i := 0; i < len(ciphertext); i++ {
		letter := ciphertext[i]
		shift := int(keyword[i%len(keyword)]) - 97
		if (int('a') <= int(letter)) && (int(letter) <= int('z')) {
			newLetter := int(letter) - shift
			if newLetter < 97 {
				newLetter += 26
			}
			plaintext += string(rune(newLetter))
		} else if (int('A') <= int(letter)) && (int(letter) <= int('Z')) {
			newLetter := int(letter) - shift
			if newLetter < 65 {
				newLetter += 26
			}
			plaintext += string(rune(newLetter))
		} else {
			plaintext += string(letter)
		}
	}

	return plaintext
}

File: homework01/vigenere.py
import string

ord_islower_ind = {True: 97, False: 65}
alphabet_length = 26


def encrypt_vigenere(plaintext: str, keyword: str) -> str:
    """
    >>> encrypt_vigenere("PYTHON", "A")
    'PYTHON'
    >>> encrypt_vigenere("python", "a")
    'python'
    >>> encrypt_vigenere("ATTACKATDAWN", "LEMON")
    'LXFOPVEFRNHR'
    """
    keyword = keyword.lower()
    ciphertext = ''
    for let_ind, letter in enumerate(plaintext):
        if letter.isalpha():
            _shift = ord(keyword[let_ind % len(keyword)]) - ord_islower_ind[True]
            ciphertext += chr(
                (ord(letter) - ord_islower_ind[letter.islower()] + _shift) % alphabet_length + ord_islower_ind[
                    letter.islower()])
        else:
            ciphertext += letter
    return ciphertext


def decrypt_vigenere(ciphertext: str, keyword: str) -> str:
    """
    >>> decrypt_vigenere("PYTHON", "A")
    'PYTHON'
    >>> decrypt_vigenere("python", "a")
    'python'
    >>> decrypt_vigenere("LXFOPVEFRNHR", "LEMON")
    'ATTACKATDAWN'
    """
    keyword = keyword.lower()
    plaintext = ''
    for let_ind, letter in enumerate(ciphertext):
        if letter.isalpha():
            _shift = ord(keyword[let_ind % len(keyword)]) - ord_islower_ind[True]
            plaintext += chr(
                (ord(letter) - ord_islower_ind[letter.islower()] - _shift) % alphabet_length + ord_islower_ind[
                    letter.islower()])
        else:
            plaintext += letter
    return plaintext


if __name__ == '__main__':
    original = string.ascii_letters
    key = 'ACEHKSPZKYREMBZQQSL'
    encrypted = encrypt_vigenere(original, key)
    decrypted = decrypt_vigenere(encrypted, key)

    print(original)
    print(encrypted)
    print(decrypted)
    print(f'Decrypted is equal to orig: {original == decrypted}')
